### PR TITLE
PERF: put BlockManager constructor in cython

### PR DESCRIPTION
**Pull request by jbrockmendel**
No idea why mypy started complaining about the io.parsers stuff

### CODE REVIEW

**Comment by jreback** [comment]
can you merge master

**Comment by jreback** [comment]
any perf on this vs master (now that the ndarray stuff is merged)? (ideally put the relevant timeits in the cython code, that is a great idea generally)

**Comment by jbrockmendel** [comment]
> any perf on this vs master (now that the ndarray stuff is merged)? (ideally put the relevant timeits in the cython code, that is a great idea generally)

On the [benchmark](https://github.com/pandas-dev/pandas/pull/40171#issuecomment-790219422) I've been using for these im seeing about a 4% gain (though the worst performance had a 0.2% slowdown).  The big win I'm expecting comes in the next step, which puts BlockManager.get_slice in cython, which allows us to chain together several no-longer-python calls

**Comment by jbrockmendel** [comment]
rebased + green

**Comment by jbrockmendel** [comment]
Just measured the branch after this (that implements BlockManager.get_slice in cython) and that gets a 15% improvement.

File: pandas/_libs/reduction.pyx [hunk 1]
```diff
@@ -489,6 +489,6 @@ cdef class BlockSlider:
         Ensure that we have the original blocks, blknos, and blklocs.
         """
         mgr = self.dummy._mgr
-        mgr.blocks = self.blocks
+        mgr.blocks = tuple(self.blocks)
```
Comment (jreback) [inline pandas/_libs/reduction.pyx:492]: this reaching in to internal stuff is a bit sketchy....

File: pandas/core/internals/managers.py [hunk 2]
```diff
@@ -1602,6 +1521,45 @@ def from_array(cls, array: ArrayLike, index: Index) -> SingleBlockManager:
         block = new_block(array, placement=slice(0, len(index)), ndim=1)
         return cls(block, index)
 
+    def __getstate__(self):
```
Comment (jreback) [inline pandas/core/internals/managers.py:1524]: need for compat? can you remove some of this?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/reduction.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/reduction.pyx:492]: yes it is.  hopefully we can optimize the non-libreduction path enough to just rip this out

File: pandas/core/internals/managers.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/internals/managers.py:1524]: this is basically the version that is in BaseBlockManager now, moved down to SingleBlockManager bc it is no longer used by BlockManager.  i do hope to get rid of some of this before long, dont have a good way to do so immediately

Comment (jbrockmendel) [reply pandas/core/internals/managers.py:1524]: (in case it was unclear, i dont expect to get rid of this for this PR)

**Review by jreback — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
> On the benchmark I've been using for these im seeing about a 4% gain

As I have mentioned before (eg on the mailing list, in https://github.com/pandas-dev/pandas/issues/40263), this benchmark is basically useless / unrealistic. 
I think you should benchmark with an actual applied function to check wether it is worth moving things to cython for this.

**Pull request merged**